### ISSUE TITLE
IfThenElse doesn't work if the food is not enumerable type 

### ISSUE THREAD

**Issue by beiciye-h-20230331** [rjsf-team/react-jsonschema-form#3780]
### Prerequisites

- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)

### What theme are you using?

core

### What is your question?

The below schema does not work anymore,  is there any doc to explain the IfThenElse process?

```
{
  "type": "object",
  "properties": {
    "animal": {
      "enum": [
        "Cat",
        "Fish"
      ]
    }
  },
  "allOf": [
    {
      "if": {
        "properties": {
          "animal": {
            "const": "Cat"
          }
        }
      },
      "then": {
        "properties": {
          "food": {
            "type": "string"
          }
        },
        "required": [
          "food"
        ]
      }
    },
    {
      "if": {
        "properties": {
          "animal": {
            "const": "Fish"
          }
        }
      },
      "then": {
        "properties": {
          "food": {
            "type": "string"
          },
          "water": {
            "type": "string",
            "enum": [
              "lake",
              "sea"
            ]
          }
        },
        "required": [
          "food",
          "water"
        ]
      }
    },
    {
      "required": [
        "animal"
      ]
    }
  ]
}
```

**Comment by nickgros** [comment]
@beiciye-h-20230331 Sorry, it's not clear what you're trying to accomplish with your schema. It would help if you could provide expected behavior.

In general, you can learn about JSON Schema's If-Then-Else logic in [Understanding JSON Schema](https://json-schema.org/understanding-json-schema/reference/conditionals.html?highlight=else#if-then-else).

A common issue is not realizing that the branch evaluation relies on the validation result of the `"if"` block. Most often, schema authors will forget to include `"required"` in their `"if"` schema to assert that a field is required. I'm not sure if that's relevant to your issue.

**Comment by nickgros** [comment]
@beiciye-h-20230331 You still haven't specified what behavior you expect.

I am assuming that you are expecting the `food` and `water` properties to not appear when `animal` is undefined.

This is because the schema
```json
{
  "properties": {
    "animal": {
      "const": "Cat"
    }
  }
}
```

actually evaluates to true against the data `{}`. See here: https://www.jsonschemavalidator.net/s/nLczXDVW

To ensure your `"if"` schemas don't evaluate to true, you could add a `"required"` property to them. [See this playground example](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6e30sInNjaGVtYSI6eyJ0eXBlIjoib2JqZWN0IiwicHJvcGVydGllcyI6eyJhbmltYWwiOnsiZW51bSI6WyJDYXQiLCJGaXNoIl19fSwiYWxsT2YiOlt7ImlmIjp7InByb3BlcnRpZXMiOnsiYW5pbWFsIjp7ImNvbnN0IjoiQ2F0In19LCJyZXF1aXJlZCI6WyJhbmltYWwiXX0sInRoZW4iOnsicHJvcGVydGllcyI6eyJmb29kIjp7InR5cGUiOiJzdHJpbmcifX0sInJlcXVpcmVkIjpbImZvb2QiXX19LHsiaWYiOnsicHJvcGVydGllcyI6eyJhbmltYWwiOnsiY29uc3QiOiJGaXNoIn19LCJyZXF1aXJlZCI6WyJhbmltYWwiXX0sInRoZW4iOnsicHJvcGVydGllcyI6eyJmb29kIjp7InR5cGUiOiJzdHJpbmcifSwid2F0ZXIiOnsidHlwZSI6InN0cmluZyIsImVudW0iOlsibGFrZSIsInNlYSJdfX0sInJlcXVpcmVkIjpbImZvb2QiLCJ3YXRlciJdfX0seyJyZXF1aXJlZCI6WyJhbmltYWwiXX1dfSwidWlTY2hlbWEiOnt9LCJ0aGVtZSI6ImRlZmF1bHQiLCJsaXZlU2V0dGluZ3MiOnt9fQ==).

**Comment by beiciye-h-20230331** [comment]
Yes, it's due to the `"required" ` missing the properties.    The issue has been fixed.  Thanks pretty much
